### PR TITLE
Support NNBD (Null safety) for generated code of mobx_codegen

### PR DESCRIPTION
**Pull request by fzyzcjy**
Hi thanks for the lib! I try to update mobx_codegen, such that the code that it *generates* will support NNBD.

Related: https://github.com/mobxjs/mobx.dart/issues/598

Also related: https://github.com/mobxjs/mobx.dart/issues/592

An example:

```
class Cls<T> extends _Cls<T> with _$Cls {
  Cls();
}

abstract class _Cls<T> with Store {
  @observable
  List<T>? genericArr;

  @observable
  List<async.Future> futureList;

  @observable
  List<async.Future<double?>>? moreList;

  // -------------------------------------------------------

  _Cls({
    this.genericArr,
    this.futureList,
    this.moreList,
  });
}
```

yields:

```
mixin _$Cls<T> on _Cls<T>, Store {
  final _$genericArrAtom = Atom(name: '_Cls.genericArr');

  @override
  List<T>? get genericArr {
    _$genericArrAtom.reportRead();
    return super.genericArr;
  }

  @override
  set genericArr(List<T>? value) {
    _$genericArrAtom.reportWrite(value, super.genericArr, () {
      super.genericArr = value;
    });
  }

  final _$futureListAtom = Atom(name: '_Cls.futureList');

  @override
  List<Future<dynamic>> get futureList {
    _$futureListAtom.reportRead();
    return super.futureList;
  }

  @override
  set futureList(List<Future<dynamic>> value) {
    _$futureListAtom.reportWrite(value, super.futureList, () {
      super.futureList = value;
    });
  }

  final _$moreListAtom = Atom(name: '_Cls.moreList');

  @override
  List<Future<double?>>? get moreList {
    _$moreListAtom.reportRead();
    return super.moreList;
  }

  @override
  set moreList(List<Future<double?>>? value) {
    _$moreListAtom.reportWrite(value, super.moreList, () {
      super.moreList = value;
    });
  }

  @override
  String toString() {
    return '''
genericArr: ${genericArr},
futureList: ${futureList},
moreList: ${moreList}
    ''';
  }
}
```

### CODE REVIEW

**Comment by VB10** [comment]
We waiting this pr, thanks u @fzyzcjy

**Comment by pavanpodila** [comment]
@fzyzcjy while you are at it, can you also see why the tests are not giving enough coverage for `mobx_codegen`?

File: mobx_codegen/pubspec.yaml [hunk 1]
```diff
@@ -12,7 +12,7 @@ dependencies:
   build: ^1.1.4
   build_resolvers: ^1.3.2
   meta: ^1.1.0
-  mobx: ^1.1.0
+  mobx: ">=1.1.0 <3.0.0"
```
Comment (pavanpodila) [inline mobx_codegen/pubspec.yaml:15]: You should specifically target the null-safety release for this

File: mobx_codegen/lib/src/type_names.dart [hunk 2]
```diff
@@ -78,25 +78,28 @@ class LibraryScopedNameFinder {
...
-        // Some types don't have associated elements, like void
-        typeElement == null ||
-            // This is a bare type param, like "T"
-            type is TypeParameterType) {
-      // TODO(pavanpodila): Once we migrate to NNBD, change the flag to `true`
-      return type.getDisplayString(withNullability: false);
-    }
-
-    return _getNamedElementTypeName(typeElement, type);
+    return type.getDisplayString(withNullability: true);
```
Comment (pavanpodila) [inline mobx_codegen/lib/src/type_names.dart:81]: Are all the tests working for this? We need to add type lookup tests that include nullable suffix

**Comment by fzyzcjy** [comment]
@pavanpodila Where can I see it?

> Merging #631 (2e42295) into master (5f85645) will increase coverage by 15.13%.

seems to say this PR make code coverage better

**Review by pavanpodila — CHANGES_REQUESTED** [review]

**Comment by pavanpodila** [comment]
> @pavanpodila Where can I see it?
> 
> > Merging #631 ([2e42295](https://github.com/mobxjs/mobx.dart/commit/2e422956f74c28774c59972ddec8d8e3349593aa)) into master ([5f85645](https://github.com/mobxjs/mobx.dart/commit/5f85645f09ce34d33d1d8edc495aeed6819017c7)) will increase coverage by 15.13%.
> 
> seems to say this PR make code coverage better

Ya that's true. It might be the upgrades to some packages. I know the code_coverage was giving some issues in the past

**Comment by pavanpodila** [comment]
Also can you PR this on the `null_safety` branch ? That also has changes for the `mobx` and `flutter_mobx` packages

File: mobx_codegen/pubspec.yaml (hunk 1 above)
Comment (fzyzcjy) [reply mobx_codegen/pubspec.yaml]: done

File: mobx_codegen/lib/src/type_names.dart (hunk 2 above)
Comment (fzyzcjy) [reply mobx_codegen/lib/src/type_names.dart:81]: Yeah I need to add a test

**Comment by fzyzcjy** [comment]
> Also can you PR this on the `null_safety` branch ? That also has changes for the `mobx` and `flutter_mobx` packages

Ah I did not notice that branch... so how should I do it now?

File: mobx_codegen/lib/src/type_names.dart (hunk 2 above)
Comment (fzyzcjy) [reply mobx_codegen/lib/src/type_names.dart]: @pavanpodila tests are added and passed

**Comment by fzyzcjy** [comment]
@pavanpodila tests are added and passed

File: mobx_codegen/test/data/valid_import_prefixed_output.dart [hunk 3]
```diff
@@ -1,10 +1,17 @@
 mixin _$User<T extends io.Process> on UserBase<T>, Store {
-  Computed<io.File> _$biographyNotesComputed;
+  Computed<io.File>? _$biographyNotesComputed;
```
Comment (pavanpodila) [inline mobx_codegen/test/data/valid_import_prefixed_output.dart:2]: In general we should have a plain, nullable and then inner nullable types as in:

- `io.File`
- `io.File?`
- `List<io.File>?`
- `List<io.File?>`
- `List<io.File?>?`

**Review by pavanpodila — CHANGES_REQUESTED** [review]
Looks good overall. In terms of type variations, pls see the comment below

File: mobx_codegen/test/data/valid_import_prefixed_output.dart (hunk 3 above)
Comment (fzyzcjy) [reply mobx_codegen/test/data/valid_import_prefixed_output.dart:2]: done

File: mobx_codegen/test/data/valid_generic_store_output.dart [hunk 4]
```diff
@@ -1,31 +1,61 @@
...
+  set values1(List<T> value) {
+    _$values1Atom.reportWrite(value, super.values1, () {
+      super.values1 = value;
+    });
+  }
+
+  final _$values2Atom = Atom(name: '_Item.values2');
+
+  @override
+  List<T> get values2 {
```
[the remark this answers is not in the document]
Comment (fzyzcjy) [reply mobx_codegen/test/data/valid_generic_store_output.dart:50]: done

**Comment by pavanpodila** [comment]
can you get the latest from `null_safety` branch...made a bunch of analysis error fixes so the Github Actions will pass :-)

**Comment by fzyzcjy** [comment]
@pavanpodila done

File: docs/docs/examples/github/index.mdx [hunk 5]
```diff
@@ -17,15 +17,15 @@ The first step is to define the reactive (aka _observable_) state of your UI. Th
 In this case, we want a simple UI where we accept the _github-ID_ as input. We use that to fetch the user's repositories. This call will be made using the <PubBadge name="github" /> package, the results for which are stored as a `List<Repository>`. The `GithubStore` class so far looks like so:
 
 ```dart
```
Comment (pavanpodila) [inline docs/docs/examples/github/index.mdx:19]: I think all the files from `docs` should not be part of this PR. Possibly because you checked out from the master branch first and then switched to null_safety ? Same goes for the README.md and other files not in `mobx_codegen`

**Review by pavanpodila — CHANGES_REQUESTED** [review]

File: docs/docs/examples/github/index.mdx (hunk 5 above)
Comment (fzyzcjy) [reply docs/docs/examples/github/index.mdx:19]: ok

Comment (fzyzcjy) [reply docs/docs/examples/github/index.mdx:19]: so I just delete all files except in mobx_codegen?

Comment (fzyzcjy) [reply docs/docs/examples/github/index.mdx:19]: done

Comment (pavanpodila) [reply docs/docs/examples/github/index.mdx]: > so I just delete all files except in mobx_codegen?

No, don't delete, else it will be a deletion when merged onto master. I think you need to revert all changes outside of mobx_codegen or possibly do a fresh checkout from null-safety and then merge changes from the mobx_codegen folder ?

**Comment by pavanpodila** [comment]
The file changes are now showing as 266 in number ! yikes!! Can you try a fresh checkout from the null_safety branch and then merge your changes into that ?

**Comment by fzyzcjy** [comment]
Ah..... Let me have a try

**Comment by pavanpodila** [comment]
hold on let me also fix some stuff from my end. I think there are some accidental changes into docs and other places in null_safety

**Comment by fzyzcjy** [comment]
ok so I just 

1. wait for some time until you are done
2. then checkout from mobx/null_safety branch into fzyzcjy/new_null_safety branch
3. merge things into that branch
4. make a new PR?

**Comment by pavanpodila** [comment]
OK, I'm done now and removed all conflicts with master. Please do a checkout from null_safety and then merge your changes

**Comment by fzyzcjy** [comment]
wait a bit, not done yet

**Comment by pavanpodila** [comment]
> ok so I just
> 
> 1. wait for some time until you are done
> 2. then checkout from mobx/null_safety branch into fzyzcjy/new_null_safety branch
> 3. merge things into that branch
> 4. make a new PR?
> 
> EDIT Or, I can just delete all files in my branch, and then merge, and then manually add back some.

How about a fresh branch ? Please close this PR and do a fresh one

**Comment by fzyzcjy** [comment]
@pavanpodila See https://github.com/mobxjs/mobx.dart/pull/634
